Port image_geometry to ROS2.

clalancette opened the pull request:
This is necessary so that we can do a conversion from a depth image to a point cloud.

connects to ros2/ros2#342

clalancette committed:
Port image_geometry to ROS2.

Signed-off-by: Chris Lalancette <clalancette@osrfoundation.org>

mikaelarguedas commented on image_geometry/CMakeLists.txt:
```diff
@@ -1,38 +1,54 @@
 add_library(${PROJECT_NAME} src/pinhole_camera_model.cpp src/stereo_camera_model.cpp)
 target_link_libraries(${PROJECT_NAME} ${OpenCV_LIBRARIES})
-add_dependencies(${PROJECT_NAME} ${catkin_EXPORTED_TARGETS})
+ament_target_dependencies(${PROJECT_NAME} "sensor_msgs" "OpenCV")
```
Nit: please use separate line when you add a list of dependencies.
also we usually list dependencies alphabetically

clalancette replied:
Yeah, I'll fix this up.

mikaelarguedas commented on image_geometry/CMakeLists.txt:
```diff
@@ -1,38 +1,54 @@
-if(CATKIN_ENABLE_TESTING)
-  add_subdirectory(test)
+if(BUILD_TESTING)
+#  add_subdirectory(test)
```
Is it possible to port the tests as well?

clalancette replied:
I'll look at the tests, see if I can easily port them.

mikaelarguedas commented on image_geometry/src/pinhole_camera_model.cpp:
```diff
@@ -1,5 +1,5 @@
 #include "image_geometry/pinhole_camera_model.h"
-#include <sensor_msgs/distortion_models.h>
+//#include <sensor_msgs/distortion_models.h>
```
Nit: space after `//` as specified in the developer guide

mikaelarguedas replied:
Is it possible to [port this file](https://github.com/ros/common_msgs/blob/jade-devel/sensor_msgs/include/sensor_msgs/distortion_models.h) rather than replacing everything with string litterals? it looks like low effort and would avoid undoing it in the future

clalancette replied:
Yeah, I've now done this in ros2/common_interfaces#31

mikaelarguedas commented on image_geometry/package.xml:
```diff
@@ -13,17 +13,21 @@
 
   <exec_depend>opencv3</exec_depend>
   <build_export_depend>sensor_msgs</build_export_depend>
+  <build_export_depend>opencv3</build_export_depend>
```
Nit: alphabetical

mikaelarguedas commented on image_geometry/include/image_geometry/pinhole_camera_model.h:
```diff
@@ -1,10 +1,11 @@
 #include <opencv2/core/core.hpp>
 #include <opencv2/imgproc/imgproc.hpp>
 #include <opencv2/calib3d/calib3d.hpp>
+#include <boost/shared_ptr.hpp>
```
can you use `std::shared_ptr` given that that's what is used everywhere in the codebase? seems overkill for this package to depend on boost only for shared pointers

clalancette replied:
This is another place where this already exists, I just added the header include to make sure it was there.  I'll open a PR for this in the ROS1 version of the package, and then we can inherit it.

clalancette replied:
Opened the PR here: ros-perception/vision_opencv#165

mikaelarguedas commented on image_geometry/package.xml:
```diff
@@ -13,17 +13,21 @@
+  <buildtool_depend>ament_cmake</buildtool_depend>
 
+  <build_depend>boost</build_depend>
   <build_depend>opencv3</build_depend>
```
can you use "depend" for opencv3? (it is equivalent to build_depend + exec_depend + build_export_depend)

clalancette replied:
Yeah, I did this (also for sensor_msgs).

mikaelarguedas reviewed (commented):
Several comments I think should be addressed before considering merging this.

clalancette committed:
Use the shared distortion_models.hpp.

Also, various other cleanup from review.

Signed-off-by: Chris Lalancette <clalancette@osrfoundation.org>

clalancette committed:
Make opencv3 and sensor_msgs a depend.

Signed-off-by: Chris Lalancette <clalancette@osrfoundation.org>

clalancette committed:
Make the tests work.

Signed-off-by: Chris Lalancette <clalancette@osrfoundation.org>

mikaelarguedas commented on image_geometry/src/pinhole_camera_model.cpp:
```diff
@@ -1,5 +1,5 @@
 #include "image_geometry/pinhole_camera_model.h"
-#include <sensor_msgs/distortion_models.h>
+#include <sensor_msgs/distortion_models.hpp>
```
is that something we are planning on submitting upstream (the file renaming I mean)?

clalancette replied:
I'm honestly not sure.  If we change it in the ROS1 package, then we'll break any downstream includers of it.  I don't think that is desirable, though it isn't the end of the world.  Also, I don't think having it be a different suffix in the ROS2 version is too confusing, and we won't ever try to be directly merging this file from upstream.  Thus, my preference would be to leave the ROS1 version as .h , and make the ROS2 version .hpp.  Thoughts?

mikaelarguedas replied:
yeah it's just that anybody wanting to migrate a package using this will have to have this extra migration step. I don't think it's the end of the world either, just trying to see how many extra steps we can avoid end users to have to go through. Same logic made me ask you to match the ROS1 folder structure for all the sensor_msgs utilities.

I don't feel very strongly about one or the other I just wanted to know what you think about it

clalancette committed:
Remove use of boost from image_geometry.

This has also been proposed in upstream ros-perception/vision_opencv
for ROS1.

Signed-off-by: Chris Lalancette <clalancette@osrfoundation.org>

clalancette committed:
Make sure to properly initialize the CameraInfo structure.

sensor_msgs::msg::CameraInfo is one of the common_interfaces
generated messages.  As such, it has a default constructor
that is empty and does no initialization.  This is important
to note, because it means that constructs such as:

class Foo
{
  protected:
    sensor_msgs::msg::CameraInfo cam_info_{};
};

Will *not* initialize all member variables of CameraInfo objects
to their default.  Thus, the PinholeCameraModel tests were
failing randomly because they were using random stack data for
their cam_info_ state.  Rectify this in PinholeCameraModel by
explicitly filling in all zeros for the CameraInfo objects.

Signed-off-by: Chris Lalancette <clalancette@osrfoundation.org>

clalancette committed:
Fix image_geometry tests.

Similar to the last commit, CameraInfo objects don't get fully
initialized to 0 (or default) when constructing.  Thus, make
sure we fill in all fields of a CameraInfo object before passing
it down into PinholeCameraModel.  Also, make another test
use this initialized version of the object, instead of a quite
new one.

Signed-off-by: Chris Lalancette <clalancette@osrfoundation.org>

clalancette commented:
CI:
windows: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=2626)](http://ci.ros2.org/job/ci_windows/2626/)
linux: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=2513)](http://ci.ros2.org/job/ci_linux/2513/)
linux aarch64: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=105)](http://ci.ros2.org/job/ci_linux-aarch64/105/)
osx: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=1982)](http://ci.ros2.org/job/ci_osx/1982/)

clalancette committed:
Fix a bunch of warnings generated by MSVC.

Signed-off-by: Chris Lalancette <clalancette@osrfoundation.org>

clalancette committed:
Fix the MSVC warnings in a different way.

Signed-off-by: Chris Lalancette <clalancette@osrfoundation.org>

clalancette commented:
All right, I finally got this to pass on all Windows/Linux/OSX.  This is ready for another review.

wjwwood commented on image_geometry/src/pinhole_camera_model.cpp:
```diff
@@ -56,11 +72,11 @@ bool update(const T& new_val, T& my_val)
 bool updateMat(const MatT& new_mat, MatT& my_mat, cv::Mat_<double>& cv_mat, int rows, int cols)
 {
-  if ((my_mat == new_mat) && (my_mat.size() == cv_mat.rows*cv_mat.cols))
+  if ((my_mat == new_mat) && (my_mat.size() == (unsigned)(cv_mat.rows*cv_mat.cols)))
```
Normally we do C++ style casts. If we're aiming to put this in `ros2.repos` then I'd ask you to investigate if this would be better done with a C++ style cast. I'm not quite sure what the right C++ alternative would be, I guess a `reinterpret_cast`, but I don't know if you can just put `unsigned` when you do that. Basically I think you'd need to know the size of the multiplication result too, and include that with the `unsigned`. Maybe there's no good C++ way to do this.

wjwwood replied:
Looking at `cv::Mat`, I think `static_cast<unsigned int>` would work, but I'm not 100% sure.

clalancette replied:
You are right, we should just use static_cast<unsigned int> here.  I'm pretty sure that will work, but I'll give it a try and see what happens.

wjwwood reviewed (approved):
lgtm, other than a minor comment.

clalancette committed:
Use a static_cast<unsigned int> instead of C-style cast.

Signed-off-by: Chris Lalancette <clalancette@osrfoundation.org>

clalancette commented:
One more CI, with the static_cast change:

CI:
windows: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=2635)](http://ci.ros2.org/job/ci_windows/2635/)
linux: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=2529)](http://ci.ros2.org/job/ci_linux/2529/)
linux aarch64: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=133)](http://ci.ros2.org/job/ci_linux-aarch64/133/)
osx: [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=1989)](http://ci.ros2.org/job/ci_osx/1989/)

clalancette commented:
All right, CI passed, merging this up.

mikaelarguedas commented on image_geometry/CMakeLists.txt:
```diff
@@ -1,38 +1,43 @@
-catkin_python_setup()
+find_package(OpenCV REQUIRED)
+find_package(sensor_msgs REQUIRED)
+find_package(Boost REQUIRED)
```
@clalancette If you got rid of the boost calls, the find_package and the dependency in the package.xml should be removed. Keeping those causes the jobs running without boost to fail: http://ci.ros2.org/job/ci_linux/2549/consoleFull

mikaelarguedas replied:
fixed in https://github.com/ros2/vision_opencv/issues/4
